src/config.js

```diff
@@ -14,40 +14,62 @@ type ApplyConfigToArgvParams = {|
   argv: Object,
   configObject: Object,
   defaultValues: Object,
+  subCommandDefaultValues: Object,
   configFileName: string,
+  commandExecuted?: string,
 |};
 
 export function applyConfigToArgv({
   argv,
   configObject,
   defaultValues = {},
+  subCommandDefaultValues,
   configFileName,
+  commandExecuted,
 }: ApplyConfigToArgvParams): Object {
-  const newArgv = {...argv};
+  let newArgv = {...argv};
   for (const option in configObject) {
+
+    validateOption(option);
+    if (option === commandExecuted) {
+      newArgv = applyConfigToArgv({
+        argv,
+        configObject: configObject[option],
+        defaultValues: subCommandDefaultValues,
+        subCommandDefaultValues,
+        configFileName});
+      continue;
+    }
+
+
     // we assume the value was set on the CLI if the default value is
     // not the same as that on the argv object as there is a very rare chance
     // of this happening
-    if (camelCase(option) !== option) {
-      throw new UsageError(`The config option "${option}" must be ` +
-        `specified in camel case: "${camelCase(option)}"`);
-    }
     const wasValueSetOnCLI = typeof(argv[option]) !== 'undefined' &&
       (argv[option] !== defaultValues[option]);
     if (wasValueSetOnCLI) {
       log.debug(`Favoring CLI: ${option}=${argv[option]} over ` +
         `configuration: ${option}=${configObject[option]}`);
       continue;
     }
+
     if (!argv.hasOwnProperty(decamelize(option, '-'))) {
       throw new UsageError(`The config file at ${configFileName} specified ` +
         `an unknown option: "${option}"`);
     }
+
     newArgv[option] = configObject[option];
   }
   return newArgv;
 }
 
+function validateOption(option: string) {
```

Comment:
by the way, try to avoid what's referred to as [premature optimization](https://shreevatsa.wordpress.com/2008/05/16/premature-optimization-is-the-root-of-all-evil/). It's very tempting and I have to fight it a lot. Only optimize what you need to optimize now.